keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,615 @@
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the
+config, but one authorizer each for other types. The ordering of this chain will
+be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
+`subjectAccessReviewVersion` accept values corresponding to flags
+`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
+`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
+respectively.
+
+This diagram shows how the configuration flags map to configuration file keys
+for a webhook:
+![](https://i.imgur.com/PJSmgOq.jpg)
+
+The `onError` will be a required field which allow users to specify whether
+or not request is denied if the webhook errors out or is unreachable. This allows
+`SubjectAccessReviews` to be more definitive.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The new structure introduces a configurable timeout that allows administrators to
+define a timeout for webhooks in case they are unreachable.
+
+> Note: ABAC will not be supported by the new file format as we discourage the usage of this mode.
+
+The code path for enabling the above will only be triggered if the feature flag will
+be enabled until the time the feature flag is removed and configuring authorizer
+through a file becomes GA.
+
+> TODO: More details on the implementation will be charted once the goals, non-goal and proposal is approved.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+- `mode` {RBAC, Node, Webhook}
+- `decision` {Allow, Deny, NoOpinion}
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track
+total invocation counts across the following labels.
+
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+3. `apiserver_authorization_step_webhook_duration_seconds`
+
+This metric will track the average latency
+
+Labels {along with possible values}:
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+4. `apiserver_authorization_step_webhook_error_total`
+
+This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.
+
+Labels {along with possible values}:
+
+- server
+- `code` {4xx, 5xx}
+- `decision` {Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+
+### Test Plan
```

Comment:
I think we should also benchmark the cost (time/memory/allocs) of some representative CEL matchConditions:
- scoping to a particular namespace
- scoping to a particular group/resource
- excluding a particular set of users
- excluding a particular set of groups